dissect/cstruct/compiler.py

```diff
@@ -1,408 +1,445 @@
+from typing import Union as UnionHint
 
 from dissect.cstruct.bitbuffer import BitBuffer
-from dissect.cstruct.expression import Expression
 from dissect.cstruct.types import (
     Array,
-    BytesInteger,
-    CharType,
-    Enum,
-    EnumInstance,
-    Field,
+    ArrayMetaType,
+    Char,
+    CharArray,
     Flag,
-    FlagInstance,
-    Instance,
-    PackedType,
+    Int,
+    MetaType,
+    Packed,
     Pointer,
-    PointerInstance,
     Structure,
     Union,
-    WcharType,
+    Void,
+    Wchar,
+    WcharArray,
 )
+from dissect.cstruct.types.packed import _struct
 
 if TYPE_CHECKING:
-    from dissect.cstruct import cstruct
+    from dissect.cstruct.cstruct import cstruct
+    from dissect.cstruct.types.structure import Field
+
+SUPPORTED_TYPES = (
+    Array,
+    Char,
+    CharArray,
+    Enum,
+    Flag,
+    Int,
+    Packed,
+    Pointer,
+    Structure,
+    Void,
+    Wchar,
+    WcharArray,
+)
+
+log = logging.getLogger(__name__)
+
+python_compile = compile
+
+
+def compile(structure: type[Structure]) -> type[Structure]:
+    return Compiler(structure.cs).compile(structure)
 
 
 class Compiler:
-    """Compiler for cstruct structures. Creates somewhat optimized parsing code."""
-
-    TYPES = (
-        Structure,
-        Pointer,
-        Enum,
-        Flag,
-        Array,
-        PackedType,
-        CharType,
-        WcharType,
-        BytesInteger,
-    )
-
-    COMPILE_TEMPLATE = """
-class {name}(Structure):
-    def __init__(self, cstruct, structure, source=None):
-        self.structure = structure
-        self.source = source
-        super().__init__(cstruct, structure.name, structure.fields, anonymous=structure.anonymous)
-
-    def _read(self, stream, context=None):
-        r = OrderedDict()
-        sizes = {{}}
-        bitreader = BitBuffer(stream, self.cstruct.endian)
-
-{read_code}
-
-        return Instance(self, r, sizes)
-
-    def add_field(self, name, type_, offset=None):
-        raise NotImplementedError("Can't add fields to a compiled structure")
-
-    def __repr__(self):
-        return '<Structure {name} +compiled>'
-"""
-
-    def __init__(self, cstruct: cstruct):
-        self.cstruct = cstruct
-
-    def compile(self, structure: Structure) -> Structure:
-        if isinstance(structure, Union) or structure.align:
-            return structure
+    def __init__(self, cs: cstruct):
+        self.cs = cs
 
-        structure_name = structure.name
+    def compile(self, structure: type[Structure]) -> type[Structure]:
+        if issubclass(structure, Union):
+            return structure
 
         try:
-            # Generate struct class based on provided structure type
-            source = self.gen_struct_class(structure_name, structure)
-        except TypeError:
-            return structure
+            structure._read = self.compile_read(structure.fields, structure.__name__, structure.align)
+            structure.__compiled__ = True
+        except Exception as e:
```

Comment:
Maybe use a more descriptive variable here? (i.e. compile_error)

Reply:
I don't really see a reason to do so, it's the only error here and it's only used for logging purposes.